pkg/satellite/stream/tcpproxy.go

```diff
@@ -0,0 +1,142 @@
+package stream
+
+import (
+	"golang.org/x/net/netutil"
+	"io"
+	"log"
+	"net"
+	"sync"
+	"time"
+)
+
+const (
+	MAX_CONNECTION   = 1
+	RECV_BUFFER_SIZE = 1024 * 1024
+	RECV_TIMEOUT     = 500 * time.Millisecond
+)
+
+type TCPProxy interface {
+	io.Closer
+}
+
+type tcpProxy struct {
+	listener        net.Listener
+	listenCloseChan chan struct{}
+	conn            net.Conn
+	stream          SatelliteStream
+
+	recvBuf       []byte
+	recvCloseChan chan struct{}
+
+	sendChan      chan []byte
+	sendCloseChan chan struct{}
+
+	closeWg sync.WaitGroup
+}
+
+func StartTCPProxy(addr string, satelliteId string) (TCPProxy, error) {
+	l, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	l = netutil.LimitListener(l, MAX_CONNECTION)
+
+	sendChan := make(chan []byte)
+
+	stream, err := OpenSatelliteStream(satelliteId, sendChan)
+	if err != nil {
+		l.Close()
+		return nil, err
+	}
+
+	t := &tcpProxy{
+		listener: l,
+		stream:   stream,
+		sendChan: sendChan,
+		recvBuf:  make([]byte, RECV_BUFFER_SIZE),
+
+		listenCloseChan: make(chan struct{}),
+		sendCloseChan:   make(chan struct{}),
+		recvCloseChan:   make(chan struct{}),
+
+		closeWg: sync.WaitGroup{},
+	}
+
+	t.start()
+
+	return t, nil
+}
+
+func (t *tcpProxy) Close() error {
+	t.stream.Close()
+
+	close(t.recvCloseChan)
+	close(t.sendCloseChan)
+	close(t.listenCloseChan)
+
+	t.closeWg.Wait()
+
+	t.conn.Close()
+	t.listener.Close()
+
+	return nil
+}
+
+func (t *tcpProxy) listen() {
+	defer t.closeWg.Done()
+	for {
+		select {
+		case <-t.listenCloseChan:
+			return
+		default:
+			conn, err := t.listener.Accept()
```

Comment:
Program execution does not proceed beyond this line unless there is a new client connecting to it, which may make a message to `listenCloseChan` is never processed. I don't know what is the right way to do it but I feel something's wrong with it.